pkg/core/interpreter_impl.go

```diff
@@ -0,0 +1,270 @@
+	for k, v := range subInterpreter.GetSlots() {
+		i.Slots[k] = v
+	}
+
+	return nil
+}
+
+// ExecuteInstruction executes a single GND instruction
+func (i *InterpreterImpl) ExecuteInstruction(op *Instruction, idx int) error {
+	if op == nil {
+		return nil
+	}
+
+	// Handle subroutine calls
+	if op.IsSubroutine {
+		// Extract the base name without .gnd extension
+		baseName := strings.TrimSuffix(op.SubroutinePath, ".gnd")
+		i.LogDebug("%s %s %v", op.Opcode, op.Destination, op.Arguments)
+
+		// Resolve arguments by mapping context properties
+		resolvedArgs, err := parsers.MapContextProperties(i.Slots, op.Arguments)
+		if err != nil {
+			return err
+		}
+
+		// Store the resolved arguments in _
+		i.LogDebug("Storing resolved args %v in _ before subroutine call", resolvedArgs)
+		i.Slots["_"] = resolvedArgs
+
+		// Execute the subroutine
+		err = i.executeSubroutine(baseName)
+		if err != nil {
+			return err
+		}
+
+		// Store the result in the destination slot
+		if val, ok := i.Slots["_"]; ok {
+			i.LogDebug("Storing subroutine result %v in destination %s", val, op.Destination)
+			i.Slots[op.Destination] = val
+		} else {
+			i.LogDebug("No result in _ to store in destination %s", op.Destination)
+		}
+
+		return nil
+	}
+
+	// Log regular instruction
+	i.LogDebug("%s %s %v", op.Opcode, op.Destination, op.Arguments)
+
+	// Resolve arguments by mapping context properties
+	resolvedArgs, err := parsers.MapContextProperties(i.Slots, op.Arguments)
+	if err != nil {
+		return err
+	}
+
+	// If no arguments provided, use current value of _
+	if len(resolvedArgs) == 0 {
+		if val, ok := i.Slots["_"]; ok {
+			i.LogDebug("Using current value of _ (%v) as argument", val)
+			resolvedArgs = []interface{}{val}
+		} else {
+			i.LogDebug("No value in _, using empty string as argument")
+			resolvedArgs = []interface{}{""}
+		}
+	}
+
+	// For return primitive, handle unquoted strings
+	if op.Opcode == "/gnd/return" {
+		// Check if all arguments are unquoted strings
+		allUnquoted := true
+		strArgs := make([]string, 0, len(op.Arguments))
+		for _, arg := range op.Arguments {
+			if strArg, ok := arg.(string); ok {
+				if len(strArg) >= 2 && strArg[0] == '"' && strArg[len(strArg)-1] == '"' {
+					allUnquoted = false
+					break
+				}
+				strArgs = append(strArgs, strArg)
+			} else {
+				allUnquoted = false
+				break
+			}
+		}
+		if allUnquoted && len(strArgs) > 0 {
+			i.LogDebug("Joining unquoted strings with spaces: %v", strArgs)
+			resolvedArgs = []interface{}{strings.Join(strArgs, " ")}
+		}
+	}
+
+	prim, ok := primitive.Get(op.Opcode)
+	if !ok {
+		return fmt.Errorf("unknown opcode: %s", op.Opcode)
+	}
+
+	result, err := prim.Execute(resolvedArgs)
+	if err != nil {
+		return fmt.Errorf("%s: %v", op.Opcode, err)
+	}
+
+	i.LogDebug("primitive result: %v", result)
+
+	// Check if this is a return with exit signal
+	if resultMap, ok := result.(map[string]interface{}); ok {
+		i.LogDebug("result is a map: %v", resultMap)
```

Comment:
We need a better type for these special return types. One which can be detected, and does not just have some properties.